qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -81,32 +111,110 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
 
         Returns:
             Output dag with UnitaryGates synthesized to target basis.
+
+        Raises:
+            TranspilerError:
+                1. pulse_optimize is True but pulse optimal decomposition is not known
+                   for requested basis.
+                2. pulse_optimize is True and natural_direction is True but a preferred
+                   gate direction can't be determined from the coupling map or the
+                   relative gate lengths.
         """
+
         euler_basis = _choose_euler_basis(self._basis_gates)
         kak_gate = _choose_kak_gate(self._basis_gates)
-
         decomposer1q, decomposer2q = None, None
         if euler_basis is not None:
             decomposer1q = one_qubit_decompose.OneQubitEulerDecomposer(euler_basis)
         if kak_gate is not None:
-            decomposer2q = TwoQubitBasisDecomposer(kak_gate, euler_basis=euler_basis)
-
-        for node in dag.named_nodes("unitary"):
+            decomposer2q = TwoQubitBasisDecomposer(
+                kak_gate, euler_basis=euler_basis, pulse_optimize=self._pulse_optimize
+            )
 
+        for node in dag.named_nodes("unitary", "swap"):
```

Comment:
Why is SWAP added here?

Reply:
This is in response to https://github.com/Qiskit/qiskit-terra/pull/6581#issuecomment-866088339

Reply:
This is useful for QV but the pass is for unitary synthesis and it seems the right way to do this is have a pass that specifically synthesized swaps. Or have a preceding pass turn the swaps into unitaries if they are to be resynthesised using this pass. Eg, the usual QV flow we could collect2qblocks/consolidateblocks before

- I guess it's maybe a common enough thing that there are gates you know will be synthesized properly using this generic machinery and you don't want to go through all that fuss to make them into unitary nodes. Perhaps this pass can take a list of node types to operate on, default 'unitary'?

Reply:
Both of these routes make sense to me. I'll have to check whether in the latter case there could be name changes on intended gates that might escape the list e.g.  "_dg" was appended or the name was replaced with something generic when converting to circuit and back to instruction. Alternatively someone has added their own version of a standard gate.

Reply:
A SWAP synthesis pass sounds right to me (might be a bit of code duplication but at least it separates concerns). At a minimum we want to be able to choose intelligently between `CNOT-NOTC-CNOT` vs. `NOTC-CNOT-NOTC` in a way that is better for the direction. I think this should be done by default, without a user flag.

Then this particular synthesis that places most of the pulses outside is, yes, good for QV but maybe not so good for other circuits (it actually *increases* the number of pulses, in the hopes that the ones on the outside absorb in other gates. But if you don't have single-qubit gates that come immediately before/after the SWAP, then you shouldn't do it).

I don't think we should convert SWAP into a unitary node, that loses the structure. And you could imagine a lot of different synthesis algorithms could later be made direction-aware and pulse-optimized (e.g. state preparation).

Reply:
Would it make sense then to leave it in for now but extract it to a swap specific pass later?

Reply:
I added a list of node types for the UnitarySynthesis pass to operate on for now.